// @hyperframes/core/dist/core.types.js
// ── Shared cross-package types ──────────────────────────────────────────────
export const CANVAS_DIMENSIONS = {
    landscape: { width: 1920, height: 1080 },
    portrait: { width: 1080, height: 1920 },
    "landscape-4k": { width: 3840, height: 2160 },
    "portrait-4k": { width: 2160, height: 3840 },
};
// `Object.keys` ordering matches insertion order in `CANVAS_DIMENSIONS` on
// every supported JS engine; tests pin the order in `index.test.ts`. Reorder
// the table above with care.
export const VALID_CANVAS_RESOLUTIONS = Object.keys(CANVAS_DIMENSIONS);
const RESOLUTION_ALIASES = {
    "1080p": "landscape",
    hd: "landscape",
    "1080p-portrait": "portrait",
    "portrait-1080p": "portrait",
    "4k": "landscape-4k",
    uhd: "landscape-4k",
    "4k-portrait": "portrait-4k",
};
/**
 * Map a user-facing resolution string (canonical name or alias) to a
 * `CanvasResolution`. Returns undefined for unknown values so callers
 * can produce their own "invalid" UX (CLI exit, route validation, etc.).
 */
export function normalizeResolutionFlag(input) {
    if (!input)
        return undefined;
    const lowered = input.toLowerCase();
    if (VALID_CANVAS_RESOLUTIONS.includes(lowered)) {
        return lowered;
    }
    return RESOLUTION_ALIASES[lowered];
}
/**
 * Runtime list of every valid `CompositionVariableType`. Use this anywhere
 * a Set/array of valid type strings is needed (lint rules, validators).
 * The `satisfies` guard turns adding a new variant to the union without
 * also adding it here into a compile error.
 */
export const COMPOSITION_VARIABLE_TYPES = [
    "string",
    "number",
    "color",
    "boolean",
    "enum",
];
export function isStringVariable(v) {
    return v.type === "string";
}
export function isNumberVariable(v) {
    return v.type === "number";
}
export function isColorVariable(v) {
    return v.type === "color";
}
export function isBooleanVariable(v) {
    return v.type === "boolean";
}
export function isEnumVariable(v) {
    return v.type === "enum";
}
export function isTextElement(el) {
    return el.type === "text";
}
export function isMediaElement(el) {
    return el.type === "video" || el.type === "image" || el.type === "audio";
}
export function isCompositionElement(el) {
    return el.type === "composition";
}
export const TIMELINE_COLORS = {
    video: "#ec4899",
    image: "#3b82f6",
    text: "#06b6d4",
    audio: "#10b981",
    composition: "#f97316",
};
export const DEFAULT_DURATIONS = {
    video: 5,
    image: 5,
    text: 2,
    audio: 5,
    composition: 5,
};
export function getDefaultStageZoom(resolution) {
    const { width, height } = CANVAS_DIMENSIONS[resolution];
    return {
        scale: 1,
        focusX: width / 2,
        focusY: height / 2,
    };
}
