// @hyperframes/engine/dist/config.js
/**
 * Engine Configuration
 *
 * Typed configuration for the rendering pipeline. Replaces the PRODUCER_*
 * env var sprawl with a structured interface. Env vars still work as
 * fallbacks for backward compatibility during migration.
 */
/** Default configuration — sensible for Hyperframes compositions. */
export const DEFAULT_CONFIG = {
    fps: 30,
    quality: "standard",
    format: "jpeg",
    jpegQuality: 80,
    concurrency: "auto",
    coresPerWorker: 2.5,
    minParallelFrames: 120,
    largeRenderThreshold: 1000,
    disableGpu: false,
    browserGpuMode: "software",
    enableBrowserPool: false,
    browserTimeout: 120_000,
    protocolTimeout: 300_000,
    forceScreenshot: false,
    enableChunkedEncode: false,
    chunkSizeFrames: 360,
    enableStreamingEncode: true,
    streamingEncodeMaxDurationSeconds: 240,
    ffmpegEncodeTimeout: 600_000,
    ffmpegProcessTimeout: 300_000,
    ffmpegStreamingTimeout: 600_000,
    hdr: false,
    hdrAutoDetect: true,
    audioGain: 1,
    frameDataUriCacheLimit: 256,
    frameDataUriCacheBytesLimitMb: 1500,
    playerReadyTimeout: 45_000,
    renderReadyTimeout: 15_000,
    verifyRuntime: true,
    debug: false,
};
/**
 * Resolve configuration by merging: defaults ← env vars ← explicit overrides.
 * Env vars provide backward compatibility during migration; explicit config
 * takes precedence over everything.
 */
export function resolveConfig(overrides) {
    const env = (key) => process.env[key];
    const envNum = (key, fallback) => {
        const raw = env(key);
        if (raw === undefined || raw === "")
            return fallback;
        const n = Number(raw);
        return Number.isFinite(n) ? n : fallback;
    };
    const envBool = (key, fallback) => {
        const raw = env(key);
        if (raw === undefined)
            return fallback;
        return raw === "true";
    };
    const envBrowserGpuMode = () => {
        const raw = env("PRODUCER_BROWSER_GPU_MODE");
        if (raw === "hardware" || raw === "software" || raw === "auto")
            return raw;
        return DEFAULT_CONFIG.browserGpuMode;
    };
    // Env-var layer (backward compat)
    const fromEnv = {
        concurrency: env("PRODUCER_MAX_WORKERS") ? Number(env("PRODUCER_MAX_WORKERS")) : undefined,
        coresPerWorker: envNum("PRODUCER_CORES_PER_WORKER", DEFAULT_CONFIG.coresPerWorker),
        minParallelFrames: envNum("PRODUCER_MIN_PARALLEL_FRAMES", DEFAULT_CONFIG.minParallelFrames),
        largeRenderThreshold: envNum("PRODUCER_LARGE_RENDER_THRESHOLD", DEFAULT_CONFIG.largeRenderThreshold),
        chromePath: env("PRODUCER_HEADLESS_SHELL_PATH"),
        disableGpu: envBool("PRODUCER_DISABLE_GPU", DEFAULT_CONFIG.disableGpu),
        browserGpuMode: envBrowserGpuMode(),
        enableBrowserPool: envBool("PRODUCER_ENABLE_BROWSER_POOL", DEFAULT_CONFIG.enableBrowserPool),
        browserTimeout: envNum("PRODUCER_PUPPETEER_LAUNCH_TIMEOUT_MS", DEFAULT_CONFIG.browserTimeout),
        protocolTimeout: envNum("PRODUCER_PUPPETEER_PROTOCOL_TIMEOUT_MS", DEFAULT_CONFIG.protocolTimeout),
        expectedChromiumMajor: env("PRODUCER_EXPECTED_CHROMIUM_MAJOR")
            ? Number(env("PRODUCER_EXPECTED_CHROMIUM_MAJOR"))
            : undefined,
        forceScreenshot: envBool("PRODUCER_FORCE_SCREENSHOT", DEFAULT_CONFIG.forceScreenshot),
        enableChunkedEncode: envBool("PRODUCER_ENABLE_CHUNKED_ENCODE", DEFAULT_CONFIG.enableChunkedEncode),
        chunkSizeFrames: Math.max(120, envNum("PRODUCER_CHUNK_SIZE_FRAMES", DEFAULT_CONFIG.chunkSizeFrames)),
        enableStreamingEncode: envBool("PRODUCER_ENABLE_STREAMING_ENCODE", DEFAULT_CONFIG.enableStreamingEncode),
        streamingEncodeMaxDurationSeconds: Math.max(0, envNum("PRODUCER_STREAMING_ENCODE_MAX_DURATION_SECONDS", DEFAULT_CONFIG.streamingEncodeMaxDurationSeconds)),
        ffmpegEncodeTimeout: envNum("FFMPEG_ENCODE_TIMEOUT_MS", DEFAULT_CONFIG.ffmpegEncodeTimeout),
        ffmpegProcessTimeout: envNum("FFMPEG_PROCESS_TIMEOUT_MS", DEFAULT_CONFIG.ffmpegProcessTimeout),
        ffmpegStreamingTimeout: envNum("FFMPEG_STREAMING_TIMEOUT_MS", DEFAULT_CONFIG.ffmpegStreamingTimeout),
        hdr: (() => {
            const raw = env("PRODUCER_HDR_TRANSFER");
            if (raw === "hlg" || raw === "pq")
                return { transfer: raw };
            return false;
        })(),
        hdrAutoDetect: envBool("PRODUCER_HDR_AUTO_DETECT", DEFAULT_CONFIG.hdrAutoDetect),
        audioGain: envNum("PRODUCER_AUDIO_GAIN", DEFAULT_CONFIG.audioGain),
        frameDataUriCacheLimit: Math.max(32, envNum("PRODUCER_FRAME_DATA_URI_CACHE_LIMIT", DEFAULT_CONFIG.frameDataUriCacheLimit)),
        frameDataUriCacheBytesLimitMb: Math.max(64, envNum("PRODUCER_FRAME_DATA_URI_CACHE_BYTES_MB", DEFAULT_CONFIG.frameDataUriCacheBytesLimitMb)),
        playerReadyTimeout: envNum("PRODUCER_PLAYER_READY_TIMEOUT_MS", DEFAULT_CONFIG.playerReadyTimeout),
        renderReadyTimeout: envNum("PRODUCER_RENDER_READY_TIMEOUT_MS", DEFAULT_CONFIG.renderReadyTimeout),
        verifyRuntime: env("PRODUCER_VERIFY_HYPERFRAME_RUNTIME") !== "false",
        runtimeManifestPath: env("PRODUCER_HYPERFRAME_MANIFEST_PATH"),
        extractCacheDir: env("HYPERFRAMES_EXTRACT_CACHE_DIR"),
    };
    // Remove undefined values so they don't override defaults
    const cleanEnv = Object.fromEntries(Object.entries(fromEnv).filter(([, v]) => v !== undefined));
    return {
        ...DEFAULT_CONFIG,
        ...cleanEnv,
        ...overrides,
    };
}
