// @hyperframes/core/dist/registry/types.js
// The `enum` arrays in `packages/core/schemas/registry*.json` must match
// `ITEM_TYPES` / `FILE_TYPES` below — `types.test.ts` is the drift guard.
// ── Constants (kept in sync with JSON Schema enums) ─────────────────────────
export const ITEM_TYPES = [
    "hyperframes:example",
    "hyperframes:block",
    "hyperframes:component",
];
export const FILE_TYPES = [
    "hyperframes:composition",
    "hyperframes:asset",
    "hyperframes:snippet",
    "hyperframes:style",
    "hyperframes:timeline",
];
/**
 * Directory segment where each item type lives under a registry root — both
 * on disk (`registry/examples/…`) and in URL construction
 * (`<baseUrl>/examples/<name>/registry-item.json`). Shared so CLIs, docs
 * tooling, and codegen scripts all agree.
 */
export const ITEM_TYPE_DIRS = {
    "hyperframes:example": "examples",
    "hyperframes:block": "blocks",
    "hyperframes:component": "components",
};
const _itemTypesExhaustive = true;
const _fileTypesExhaustive = true;
void _itemTypesExhaustive;
void _fileTypesExhaustive;
// ── Type guards ─────────────────────────────────────────────────────────────
export function isExampleItem(item) {
    return item.type === "hyperframes:example";
}
export function isBlockItem(item) {
    return item.type === "hyperframes:block";
}
export function isComponentItem(item) {
    return item.type === "hyperframes:component";
}
